# 2. Utilizando Python como una calculadora

**Por Arles Rodríguez PhD**

## Identificadores y variables en python

Un *identificador* es una secuencia de símbolos que se utilizan como nombres de variables, funciones, clases y otras estructuras de los lenguajes de programación.

Los identificadores se escriben como secuencias de caracteres alfanuméricos del alfabeto inglés. En python se utiliza una notación conocida como snake_case (ó palabras en minúsculas separadas por guiones) para evitar utilizar espacios como por ejemplo: conteo_animales, nombre_usuario, velocidad, etc. Se recomienda utilizar únicamente letras del alfabeto inglés a la hora de usar identificadores.


### Palabras reservadas 

```
# Estas son las palabras reservadas de python
False      await      else       import     pass
None       break      except     in         raise
True       class      finally    is         return
and        continue   for        lambda     try
as         def        from       nonlocal   while
assert     del        global     not        with
async      elif       if         or         yield

```

Fuente:  https://docs.python.org/3/reference/lexical_analysis.html#identifiers

Son ejemplos de identificadores válidos:

```
i
x
suma
sumando1
sumando2
edad 
pais_de_nacimiento
nombre 
area_circulo

```

Los identificadores son sensibles a mayúsculas y minúsculas, esto quiere decir que por ejemplo los identificadores `dia`,  `DIA` y `Dia` harían referencia a variables distintas.



In [22]:
Dia = 'lunes'
dia = 'martes'
DIA = 'miercoles'

print('Dia={} dia={} DIA={}'.format(Dia,dia,DIA))

Dia=lunes dia=martes DIA=miercoles


## Tipos de datos escalares

Los programas manipularán diferentes tipos de datos. Cada objeto tiene un tipo de datos asociado a el. Se tienen tipos de datos escalares que corresponden a: 

- `int` representa números enteros ejemplo: 5, -5
- `float` representa números reales 3.27
- `bool` representa valores booleanos `True` y `False`
- `Nonetype` es especial y tiene un único valor `None`

es posible usar type() para ver el tipo de dato de un objeto:

In [27]:
x = 5.6
type(x)

float

In [28]:
type(5)

int

In [29]:
type(3.0)

float

In [30]:
type(None)

NoneType

In [31]:
type(True)

bool

In [46]:
city='London' # A string variable assignment.
money = 100.75 # A floating point number assignment
count=4 #An integer assignment

In [47]:
type(city)

str

In [48]:
type(money)

float

In [49]:
city = 2
type(city)

int

## Operaciones aritméticas y lógicas

Los operadores aritméticos principales son:


- Suma: +
- Resta: -
- Multiplicación: *
- División entera: //
- División real: /
- Residuo: %
- Potencia: **
- Asignación: =
- Asignación con suma: +=
- Asignación con resta: -=

A continuación se muestran algunos ejemplos de operaciones básicas:

In [50]:
#división que te va a dar como resultado un número real.
7/3


2.3333333333333335

In [20]:
#División que te va a dar como resultado un número entero
7//3


2

In [51]:
1//4

0

In [52]:
1/4

0.25

In [53]:
#Ejemplo de la operación residuo
17 % 3

2

In [54]:
8*8*8*8*8*8*8

2097152

In [55]:
8**7

2097152

In [59]:
x = 5*3
print(x)

15


In [60]:
x = x + 2
print(x)

17


In [61]:
x += 2 # x = x + 2
print(x)

19


In [62]:
y =  4 * 6
z = 5
y += z
print(y)

29


# Operadores lógicos y de Comparación:

Python posee los siguientes operadores lógicos básicos:

- ```and, or, not``` 

Y los siguientes operadores de comparación:

- ```a == b```: retorna verdadero si los valores de a y b son iguales, falso en otro caso.
- ```a != b ```: retorna verdadero si los valores son diferentes, falso en otro caso.
- ```a >= b```: retorna verdadero si a es mayor o igual a b. 
- ```a > b```: retorna verdadero si a es mayor a b. 
- ```a <= b```: retorna verdadero si a es menor o igual a b.
- ```a < b```: retorna verdadero si a es menor a b.

A continuación algunos ejemplos:

In [64]:
a = 3
b = 4

print(a == b)

False


In [65]:
x = 3
y = 5
print(x > y)

False


In [66]:
-4 <= -7

False

In [31]:
x = 5*3
y = 15

print(x == y)

True


In [68]:
Claudia_pone_metro = False
Claudia_es_buena_alcaldesa = True

print(Claudia_pone_metro and Claudia_es_buena_alcaldesa)

False


In [69]:
p = False # hoy es lunes
q = False  # dos es impar

print(not p or q)

True


# Operaciones a nivel de bits

Python también soporta operaciones a nivel de bits:
- ```|``` or lógica
- ```&``` and lógico
- ```~``` not (unario)
- ```^``` xor
- ```>>``` corrimiento en bits a derecha
- ```<<``` corrimiento en bits a izquierda
    

In [70]:
24 | 5

29

In [71]:
24 & 5

0

In [72]:
2422 >> 1

1211

In [11]:
5 & 1 

1

In [12]:
4 & 1

0

In [74]:
1 << 2

4

In [86]:
24^4 #xor

28

In [87]:
~4

-5

In [89]:
63 & ~(1 << 3)

55

## Conversión de tipos (casting)

Para realizar la conversión de tipos se utiliza el nombre del tipo de dato y paréntesis:

In [91]:
int(4.77)

4

In [8]:
float(65)

65.0

In [92]:
str(98)

'98'

## Lectura de variables desde la entrada estándar

Es posible leer valores digitados por el usuario de la siguiente forma:

In [110]:
x = input()
print("valor leido: " + x)

Jaime
valor leido: Jaime


In [111]:
nombre = input('digite su nombre:')
print("Hola " + nombre + "!!!!!")

digite su nombre:Arles
Hola Arles!!!!!


In [112]:
valor_compra = float(input('digite el valor de la compra: '))
#crear la variable porcentaje descuento y leer el valor
descuento = valor_compra*0.1 #hay un descuento del 10%
valor_a_pagar = valor_compra - descuento
print('El valor a pagar es: ' + str(valor_a_pagar))
print('Tu ahorro fue de: ' + str(descuento))

digite el valor de la compra: 10000
El valor a pagar es: 9000.0
Tu ahorro fue de: 1000.0


### Ejercicio 1

- Ejercicio 1: Utilizando el programa anterior, leer el descuento y calcular valor a pagar y ahorro.


### Ejercicio 2

Elabore un programa que Calcule el área de un rectángulo.


### Ejercicio 3: 

Elabore un programa que Calcule el área de un círculo.

In [116]:
from math import pi
print(pi)


3.141592653589793


## Funciones (Métodos) en Python 

En programación, así como en matemáticas, para las funciones definidas como $f: A \to B$, al conjunto $A$ se le denomina dominio y al conjunto $B$ como rango. A partir de estos objetos se construye el encabezado de las funciones de programación.

Sobre esta función se tiene que $f$ corresponde al nombre de la función, el conjunto $A$ corresponde al tipo de los argumentos de dicha función y el conjunto $B$ que es el rango corresponderá al valor de retorno de dicha función.

### Funciones con un parámetro de entrada: cuadrado de un número

Se definirá una función que eleve un número dado al cuadrado. Para expresar una función que calcule esta operación, en primera instancia debe pensar en el dominio y en el rango de la función $f: \mathbb{R} \to \mathbb{R}$ que define la función tiene como entrada (dominio) un número real $x$ y como salida (rango) un número real. 

La declaración de la función junto con su cuerpo quedará de la siguiente forma:

In [129]:
def cuadrado(x):
   """
   Input: x, a float
   Returns square of x
   """
   return x**2

print(cuadrado(3.3))


10.889999999999999


### Área de un círculo 

Para el desarrollo de esta función lo primero es determinar el nombre. La función se llamará $area\_circulo$ cuyo dominio es el conjunto de los números reales (para el radio) y cuyo rango pertenece al conjunto de los números reales (el valor de retorno que corresponde al área del círculo).  

Teniendo en cuenta que el algoritmo para el cálculo del área de un círculo depende del valor de su radio, entonces, el área del círculo está dada por la expresión $A_c = \pi*r^2$, donde las variables están definidas y el planteamiento matemático de la función solicitada será el siguiente:

\begin{align*}
area\_circulo: \mathbb{R} &\to \mathbb{R}\\
(r) &\mapsto 3.14159265*r*r
\end{align*}

En python se tendrá:


In [6]:
from math import pi

def area_circulo(r):
  """
  Input: r correspondiente al radio
  Returns: área del círculo
  """
  return pi*r*r

print("area:", area_circulo(5))

area: 78.53981633974483


### Funciones con más de un parámetro de entrada: área de un rectángulo

La función que calcula el área de un rectángulo es una función que tiene como parámetros de entrada el ancho y el largo del rectángulo que son valores de tipo real y retorna como salida el área del rectángulo. Lo que puede escribirse como $$area\_rectangulo: \mathbb{R} \times \mathbb{R} \to \mathbb{R}.$$

Para el calculo del área de un rectángulo es necesario conocer el largo y el ancho del rectángulo, a partir de los cuales el área del rectángulo está dada por la expresión $A_r = l * a$, donde las variables están definidas así:


Entonces, la función matemática queda definida de la siguiente forma:

\begin{align*}
area\_rectangulo: \mathbb{R} \times \mathbb{R} &\to \mathbb{R}\\
(l,a) &\mapsto l*a
\end{align*}


In [3]:
def area_rectangulo(l, a):
  """
  Input: l correspondiente al largo
         a correspondiente al ancho
  Returns: área del rectángulo
  """
  return l*a

In [2]:
largo = float(input('digite el largo del rectángunlo: '))
ancho = float(input('digite el ancho del rectángunlo: '))
print("El área del rectángulo es:", area_rectangulo(largo, ancho))

digite el largo del rectángunlo: 34
digite el ancho del rectángunlo: 54
El área del rectángulo es: 1836.0


### Ejercicio 1: 

Utilizando los métodos area_rectangulo y area_circulo, escriba un método para calcular el área del vagón que se da a continuación:

![](images/vagon.png "El vagon")

In [17]:
def area_vagon(b, a, r):
    #lalala
    #sjkdfjksa
    #akjsfjkas
    return area_rectangulo(b,a) + 2*area_circulo(r)


In [18]:
print(area_vagon(5,2,1))

16.283185307179586


### Ejercicio 2: 

Utilizando los métodos area_rectangulo y area_circulo, escriba un método para calcular el área del carrito que se da a continuación:

![](images/carrito.png "El carrito")

### Ejercicio 3

Dada la edad de una persona en días, calcule la edad de la persona en años, meses y días:
Nota: Para facilitar los cálculos, considere que todos los años tienen 365 días, los meses 30 días. Es un ejercicio para probar su razonamiento matemático básico.

Ejemplos:

<table>
    <tr>
        <td>
        Input
        </td>
        <td>
        Output
        </td>
    </tr>
    <tr>
        <td>
        30
        </td>
        <td>
        0 año(s)
        1 mes(es)
        0 día(s) 
        </td>
    </tr>   
    <tr>
        <td>
        400
        </td>
        <td>
        1 año(s)
        1 mes(es)
        5 día(s) 
        </td>
    </tr>
      <tr>
        <td>
        550
        </td>
        <td>
        1 año(s)
        6 mes(es)
        5 día(s) 
        </td>
    </tr>
</table>

Truco para VPL: Añadir en la primera línea, la siguiente línea para especificar codificación:
```# -*- coding: latin-1 -*-```

1000
2 años 9 mes(es) 0día(s)


## Referencias:

https://docs.python.org/3/reference/lexical_analysis.html#identifiers

Das, B. N. (2017). Learn Python in 7 Days. Packt Publishing Ltd.


Gomez, J, Rodriguez A y Cubides C. La ciencia de Programar. Universidad Nacional de Colombia.